### PR TITLE
Refactored LTI related classes, using interfaces and implementations from library 

### PR DESCRIPTION
**Pull request by emgolubev**
Related to: https://oat-sa.atlassian.net/browse/TR-2279

### CODE REVIEW

File: models/classes/Platform/LtiPlatform.php [hunk 1]
```diff
@@ -22,92 +22,94 @@
 
 namespace oat\taoLti\models\classes\Platform;
 
+use OAT\Library\Lti1p3Core\Platform\PlatformInterface;
+
 /**
  * LTI platform Value Object.
  */
-class LtiPlatform
+class LtiPlatform implements PlatformInterface
```
Comment (quintanilhar) [inline models/classes/Platform/LtiPlatform.php:30]: Can you help me to understand why we need this custom implementation, please? 

There is an implementation already provided by the core lib.

Comment (emgolubev) [reply models/classes/Platform/LtiPlatform.php:30]: This implementation has additional properties and I could not extend the class from the library, it has private methods

File: models/classes/ServiceProvider/LtiServiceProvider.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            ->args(
+                [
+                    service(ClientRepositoryInterface::class),
+                    service(AccessTokenRepositoryInterface::class),
+                    service(ScopeRepositoryInterface::class),
+                    param('encryptionKey')
+                ]
+            );
+    }
+}
```
Comment (poyuki) [inline models/classes/ServiceProvider/LtiServiceProvider.php:129]:
```suggestion
}

```

File: controller/Security.php [hunk 3]
```diff
@@ -80,6 +85,21 @@ public function oidcInitiation(): void
...
+    }
+
+    private function getPlatformKeyChainRepository(): PlatformKeyChainRepository
+    {
+        return $this->getServiceLocator()->get(PlatformKeyChainRepository::SERVICE_ID);
+    }
+
+    private function getAuthorizationServerFactory(): AuthorizationServerFactory
+    {
+        return $this->getServiceLocator()->getContainer()->get(AuthorizationServerFactory::class);
```
Comment (wazelin) [inline controller/Security.php:100]: I'm pretty sure these can be unified by accessing a service by its FQN.
```suggestion
        return $this->getServiceLocator()->get(PlatformKeyChainRepository::class);
    }

    private function getAuthorizationServerFactory(): AuthorizationServerFactory
    {
        return $this->getServiceLocator()->get(AuthorizationServerFactory::class);
```

Comment (wazelin) [reply controller/Security.php:100]: Perhaps, additional changes will be required in the container mock configuration as well in some of the tests.

**Review by poyuki — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`

File: manifest.php [hunk 4]
```diff
@@ -85,5 +86,8 @@
     ],
     'extra' => [
         'structures' => $extpath . 'controller' . DIRECTORY_SEPARATOR . 'structures.xml',
-    ]
+    ],
+    'containerServiceProviders' => [
```
Comment (boajer) [inline manifest.php:90]: Please ensure `core` and `generis` versions in `composer.json` matches this new feature

File: models/classes/ServiceProvider/LtiServiceProvider.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+        $services
+            ->set(JwksFetcherInterface::class, JwksFetcher::class)
+            ->public()
+            ->args(
+                [
+                    service(ItemPoolSimpleCacheAdapter::class),
+                    null,
+                    null,
+                    service(LoggerInterface::class)
```
Comment (boajer) [inline models/classes/ServiceProvider/LtiServiceProvider.php:71]:
```suggestion
                    service(LoggerService::SERVICE_ID)
```

File: models/classes/ServiceProvider/LtiServiceProvider.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+        /** @todo how to get logger? */
+        $services
+            ->set(LoggerInterface::class, Logger::class)
+            ->public()
+            ->args(
+                [
+                    'general'
+                ]
+            );
```
Comment (boajer) [inline models/classes/ServiceProvider/LtiServiceProvider.php:61]:
```suggestion      
```

File: models/classes/ServiceProvider/LtiServiceProvider.php [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+        $services
+            ->set(ClientRepositoryInterface::class, ClientRepository::class)
+            ->public()
+            ->args(
+                [
+                    service(Lti1p3RegistrationRepository::class),
+                    service(JwksFetcherInterface::class),
+                    service(LoggerInterface::class)
```
Comment (boajer) [inline models/classes/ServiceProvider/LtiServiceProvider.php:82]:
```suggestion
                    service(LoggerService::SERVICE_ID)
```

File: models/classes/ServiceProvider/LtiServiceProvider.php [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                ]
+            );
+
+        $services
+            ->set(AccessTokenRepositoryInterface::class, AccessTokenRepository::class)
+            ->public()
+            ->args(
+                [
+                    service(ItemPoolSimpleCacheAdapter::class),
+                    service(LoggerInterface::class)
```
Comment (boajer) [inline models/classes/ServiceProvider/LtiServiceProvider.php:92]:
```suggestion
                    service(LoggerService::SERVICE_ID)
```

File: models/classes/Platform/LtiPlatform.php (hunk 1 above)
Comment (emgolubev) [reply models/classes/Platform/LtiPlatform.php:30]: Let's just rename it while

File: models/classes/user/Lti1p3User.php [hunk 9]
```diff
@@ -22,13 +22,17 @@
...
+use OAT\Library\Lti1p3Core\Registration\RegistrationInterface;
 use oat\taoLti\models\classes\LtiLaunchData;
 use oat\taoLti\models\classes\LtiRoles;
 use oat\taoLti\models\classes\LtiUtils;
 use oat\taoLti\models\classes\LtiVariableMissingException;
 
 class Lti1p3User extends LtiUser
 {
+    /** @var RegistrationInterface */
+    private $registration = null;
```
[the remark this answers is not in the document]
Comment (emgolubev) [reply models/classes/user/Lti1p3User.php:34]: It will be needed for AGS, remove it while

File: manifest.php (hunk 4 above)
Comment (emgolubev) [reply manifest.php:90]: Only `generis` should be bumped as I can see

File: migrations/Version202110181634585506_taoLti.php [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    public function up(Schema $schema): void
+    {
+        $configPath = __DIR__ . '/../config/taoLti/AuthorizationServerFactory.conf.php';
+
+        if (file_exists($configPath) && unlink($configPath)) {
+            $this->addReport(Report::createInfo('AuthorizationServerFactory.conf.php has been removed.'));
+        }
+
+        $this->addReport(
+            Report::createInfo(
```
Comment (boajer) [inline migrations/Version202110181634585506_taoLti.php:33]: I Would suggest specifying a higher report level for better visibility 
```suggestion
            Report::createWarning(
```

File: models/classes/Security/DataAccess/Repository/PlatformKeyChainRepository.php [hunk 11]
```diff
@@ -41,27 +41,42 @@ class PlatformKeyChainRepository extends ConfigurableService implements KeyChain
...
 
         if (!$isPublicKeySaved || !$isPrivateKeySaved) {
             throw new ErrorException('Impossible to write LTI keys');
         }
     }
 
-    public function findAll(KeyChainQuery $query): KeyChainCollection
+    public function getDefaultKeyId(): string
+    {
+        return $this->getOption(PlatformKeyChainRepository::OPTION_DEFAULT_KEY_ID) ?? '';
```
Comment (boajer) [inline models/classes/Security/DataAccess/Repository/PlatformKeyChainRepository.php:68]:
```suggestion
        return $this->getOption(PlatformKeyChainRepository::OPTION_DEFAULT_KEY_ID, '');
```

File: composer.json [hunk 12]
```diff
@@ -58,7 +58,7 @@
         "ext-openssl": "*",
         "oat-sa/oatbox-extension-installer": "~1.1||dev-master",
         "oat-sa/lib-lti1p3-core": "^6.0.0",
-        "oat-sa/generis" : ">=14.0.0",
+        "oat-sa/generis" : ">=15.5.0",
         "oat-sa/tao-core" : ">=48.30.0"
```
Comment (boajer) [inline composer.json:62]:
```suggestion
        "oat-sa/tao-core" : ">=48.39.0"
```

**Review by boajer — APPROVED** [review]
- [ ]  New code is covered by tests (if applicable)
- [ ]  Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x]  New code is respecting code style rules
- [x]  New code is respecting best practices
- [ ]  New code is not subject to concurrency issues (if applicable)
- [ ]  Feature is working correctly on my local machine (if applicable)
- [ ]  Acceptance criteria are respected
- [x]  Pull request title and description are meaningful
- [ ]  Pull request's target is not master
- [ ]  Commits are following conventional commits
- [ ]  Commits messages are meaningful
- [ ]  Commits are atomic
- [ ]  Changelog is updated according to changes (if applicable)
- [ ]  Documentation is updated according to changes (if applicable)

**Pull request merged**